### ISSUE TITLE
hidden states for backwards

### ISSUE THREAD

**Issue by onlytailei** [ikostrikov/pytorch-a3c#15]
In the training part,
The output value and loss are saved at first.
At the end of the forward stage, they are calculated.
The problem is, at that time. the hidden states of the model are all from the last input state.
In the backward procedure, the gradients of weights are also depending on the hidden states.
But the outputs of former saved states are obviously based their own hidden states but not the hidden states right now. 
So can the variable.backward() function wok well for this situation?

**Comment by ikostrikov** [comment]
I'm confused. Can you point to the exact lines?

**Comment by onlytailei** [comment]
```
        for i in reversed(range(len(rewards))):
            R = args.gamma * R + rewards[i]
            advantage = R - values[i]
            value_loss = value_loss + 0.5 * advantage.pow(2)

            # Generalized Advantage Estimataion
            delta_t = rewards[i] + args.gamma * \
                values[i + 1].data - values[i].data
            gae = gae * args.gamma * args.tau + delta_t

            policy_loss = policy_loss - \
                log_probs[i] * Variable(gae) - 0.01 * entropies[i]

        optimizer.zero_grad()

        (policy_loss + 0.5 * value_loss).backward()
        torch.nn.utils.clip_grad_norm(model.parameters(), 40)

        ensure_shared_grads(model, shared_model)
        optimizer.step()

```

**Comment by ikostrikov** [comment]
But why the hidden states are from the last input state? PyTorch creates new variables for each step that are not overwritten.

**Comment by onlytailei** [comment]
Thank you.
I will check it.
